src/types.d.ts

```diff
@@ -0,0 +1,148 @@
+  opacityBy: null | DataEncoding;
+  xScale: null | Scale;
+  yScale: null | Scale;
+  pointConnectionColorBy: null | PointDataEncoding;
+  pointConnectionOpacityBy: null | PointDataEncoding;
+  pointConnectionSizeBy: null | PointDataEncoding;
+};
+
+type PointOptions = {
+  pointColor: Color;
+  pointColorActive: Color;
+  pointColorHover: Color;
+  pointOutlineWidth: number;
+  pointSize: number;
+  pointSizeSelected: number;
+  pointConnectionColor: Color;
+  pointConnectionColorActive: Color;
+  pointConnectionColorHover: Color;
+  pointConnectionOpacity: number;
+  pointConnectionOpacityActive: number;
+  pointConnectionSize: number;
+  pointConnectionSizeActive: number;
+  pointConnectionMaxIntPointsPerSegment: number;
+  pointConnectionTolerance: number;
+};
+
+type CameraOptions = {
+  cameraTarget: [number, number];
+  cameraDistance: number;
+  cameraRotation: number;
+  cameraView: Float32Array;
+};
+
+type LassoOptions = {
+  lassoColor: Color;
+  lassoLineWidth: number;
+  lassoMinDelay: number;
+  lassoMinDist: number;
+  lassoClearEvent: 'lassoEnd' | 'deselect';
+  lassoInitiator: boolean;
+  lassoInitiatorParentElement: HTMLElement;
+};
+
+export type Settable = {
+  backgroundColor: Color;
+  deselectOnDblClick: boolean;
+  deselectOnEscape: boolean;
+  keyMap: KeyMap;
+  mouseMode: MouseMode;
+  showPointConnections: boolean;
+  showReticle: boolean;
+  reticleColor: Color;
+  opacity: number;
+  opacityByDensityFill: number;
+  height: 'auto' | number;
+  width: 'auto' | number;
+  gamma: number;
+} & PointOptions &
+  CameraOptions &
+  LassoOptions &
+  Nullifiable;
+
+export type Properties = {
+  canvas: HTMLCanvasElement;
+  regl: import('regl').Regl;
+  syncEvents: boolean;
+  version: string;
+  lassoInitiatorElement: HTMLElement;
+  camera: Camera2D;
+  performanceMode: boolean;
+  opacityByDensityDebounceTime: number;
+} & Settable;
+
+// Options for plot.{draw, select, hover, deselect}
+
+export type ScatterPlotOptions = {
+  draw: Partial<{
+    transition: boolean;
+    transitionDuration: number;
+    transitionEasing: string;
+  }>;
+  hover: Partial<{
+    showReticleOnce: boolean;
+    preventEvent: boolean;
+  }>;
+  select: Partial<{
+    merge: boolean;
+    preventEvent: boolean;
+  }>;
+};
+
+// PubSub definitions
+
+type PubSubEvent<EventName extends string, Payload extends unknown> = {
+  [Key in EventName]: Payload;
+};
+
+type DrawPayload = Pick<Properties, 'camera' | 'xScale' | 'yScale'> & {
+  view: Properties['cameraView'];
+};
+
+type EventMap = PubSubEvent<
+  | 'init'
+  | 'destroy'
+  | 'backgroundImageReady'
+  | 'deselect'
+  | 'lassoStart'
+  | 'transitionStart'
+  | 'pointConnectionsDraw',
+  undefined
+> &
+  PubSubEvent<'view' | 'draw', DrawPayload> &
+  PubSubEvent<'lassoEnd' | 'lassoExtend', { coordinates: number[] }> &
+  PubSubEvent<'pointOver' | 'pointOut', number> &
+  PubSubEvent<'points', { points: number[][] }> &
+  PubSubEvent<'transitionEnd', import('regl').Regl>;
```

Comment:
Maybe a little confusing, but I think the easiest way to define or extend the behavior of the custom pubsub regl-scatter wraps. Let me know if you have any questions on extending this.